sync_sushi.c

```diff
@@ -1,41 +1,96 @@
 #include <pthread.h>
 #include <semaphore.h>
 #include <stdio.h>
+#include <stdlib.h>
 #include <unistd.h>
 #include "customer.h"
 
 #define BAR_SIZE 5 // number of seats in the bar
 #define CUSTOMERS_SIZE 20 // number of customers;
 
-pthread_mutex_t mutex;
-sem_t block;
+pthread_mutex_t mutex; // blocks the usage of updating customers_eating and customers_waiting
+sem_t seats; // semaphore that tracks the amount of seats
+
 int customers_eating = 0;
 int customers_waiting = 0;
 int must_wait = 0;
 
 Customer customers[CUSTOMERS_SIZE]; // collection of customers;
 pthread_t customers_threads[CUSTOMERS_SIZE]; // collection of customers;
 
-void* dinner(void* arg) {}
+void* dinner(void* arg)
+{ 
+	Customer* customer = (Customer*) arg;
+	while(1)
+	{
+		pthread_mutex_lock(&mutex); // locks mutex
+		if(must_wait) // if bar is full, customer has to wait
+		{ 
+			customers_waiting += 1;
+			customer->status = waiting;
+			
+			pthread_mutex_unlock(&mutex); // releases mutex allowing counters update
+			sem_wait(&seats); // blocks thread until a seat becomes available
+
+			customers_waiting -= 1;
+		} 
+
+		customers_eating += 1;
+		customer->status = eating;
+		must_wait = (customers_eating == BAR_SIZE);
+		if(customers_waiting && !must_wait) 
```

Comment:
In which case, the thread enters in that " if " ?

Reply:
Oh, I notice a commented wrong in line :42. A thread enters this conditional expression block when a customer can leave (it's not a party of five). This approach is noted down as _solution 2_, that says:
> ... is based on the counterintuitive notion that we cantransfer a mutex from one thread to another! In other words, one thread canacquire a lock and then another thread can release it. As longas both threadsunderstand that the lock has been transferred, there is nothing wrong with this.